proof/courier.go

```diff
@@ -849,6 +921,206 @@ func (h *HashMailCourier) SetSubscribers(
+	// service.
+	for i := 0; i < proofFile.NumProofs(); i++ {
+		transitionProof, err := proofFile.ProofAt(uint32(i))
+		if err != nil {
+			return err
+		}
+		proofAsset := transitionProof.Asset
+
+		// Construct asset leaf.
+		rpcAsset, err := taprpc.MarshalAsset(
+			ctx, &proofAsset, true, true, nil,
+		)
+		if err != nil {
+			return err
+		}
+
+		var proofBuf bytes.Buffer
+		if err := transitionProof.Encode(&proofBuf); err != nil {
+			return fmt.Errorf("error encoding proof file: %w", err)
+		}
+
+		assetLeaf := unirpc.AssetLeaf{
+			Asset:         rpcAsset,
+			IssuanceProof: proofBuf.Bytes(),
+		}
+
+		// Construct universe key.
+		outPoint := transitionProof.OutPoint()
+		assetKey := unirpc.MarshalAssetKey(
+			outPoint, proofAsset.ScriptKey.PubKey,
+		)
+		universeID := unirpc.MarshalUniverseID(proofAsset.ID())
+		universeKey := unirpc.UniverseKey{
+			Id:      universeID,
+			LeafKey: assetKey,
+		}
+
+		// Submit proof to courier.
+		_, err = c.client.InsertProof(ctx, &unirpc.AssetProof{
+			Key:       &universeKey,
+			AssetLeaf: &assetLeaf,
+		})
+		if err != nil {
+			return fmt.Errorf("error inserting proof into "+
+				"universe courier service: %w", err)
+		}
+	}
+
+	return err
+}
+
+// ReceiveProof attempts to obtain a proof file from the courier service. The
+// final proof in the target proof file is identified by the given locator.
+func (c *UniverseRpcCourier) ReceiveProof(ctx context.Context,
+	originLocator Locator) (*AnnotatedProof, error) {
+
+	// Sleep for a bit to give the sender sufficient time to transfer the
+	// proof to the proof courier service.
+	//
+	// TODO(ffranr): Replace this with the backoff procedure as found in the
+	// hashmail courier.
+	time.Sleep(5 * time.Second)
+
+	// In order to reconstruct the proof file we must collect all the
+	// transition proofs that make up the main chain of proofs. That is
+	// accomplished by iterating backwards through the main chain of proofs
+	// until we reach the genesis point (minting proof).
+
+	// We will update the locator at each iteration.
+	loc := originLocator
+
+	// revProofs is a slice of transition proofs ordered from latest to
+	// earliest (the issuance proof comes last in the slice). This ordering
+	// is a reversal of that found in the proof file.
+	var revProofs []Proof
+
+	for {
+		universeID := unirpc.MarshalUniverseID(*loc.AssetID)
+		assetKey := unirpc.MarshalAssetKey(
+			*loc.OutPoint, &loc.ScriptKey,
+		)
+		universeKey := unirpc.UniverseKey{
+			Id:      universeID,
+			LeafKey: assetKey,
+		}
+
+		resp, err := c.client.QueryProof(ctx, &universeKey)
+		if err != nil {
+			return nil, err
+		}
+
+		// Decode transition proof from query response.
+		proofBlob := resp.AssetLeaf.IssuanceProof
+		var transitionProof Proof
+		if err := transitionProof.Decode(
+			bytes.NewReader(proofBlob),
+		); err != nil {
+			return nil, err
+		}
+
+		revProofs = append(revProofs, transitionProof)
```

Comment:
Re this and above, what about the additional inputs? So the case where you have an asset, but it was created by merging 3 change UTXOs.

Reply:
Each individual (transition) `Proof` contains a field called ` AdditionalInputs []File`. The entire proof file for each additional inputs is included in that field. At least, that's my current understanding after speaking with Oli.

We noticed that this is a memory inefficiency within the universe tree (on the courier server side). Each leaf of the universe tree contains a proof and each one of those proofs may contain entire proof files for additional inputs. Those proof files may already exist in the universe as other leaves.

However, from a tapd peer sending/receiving perspective, this setup makes things simpler. I'll add an issue to talk about the potential optimisation here. It will help me get a clear picture of what's going on.

Reply:
Issue here: https://github.com/lightninglabs/taproot-assets/issues/503